File: Python/Sorting/insertion_Sort.py
# Best Case: O(n) Time Complexity | O(1) Space Complexity
# Average Case: O(n^2) Time Complexity | O(1) Space Complexity
# Worst Case: O(n^2) Time Complexity | O(1) Space Complexity

def insertion_sort(array):
    """Function to implement insertion sort algorithm"""
    for i in range(1, len(array)):
        # Looping through array once
        value_sort = array[i] # Current number
        while array[i - 1] > value_sort and i > 0:
            swap(i, i - 1, array) # Swaping to insert current number in sorted sublist
            i = i - 1
    return array

def swap(i, j, array):
    """Function to swap two elements in the same array"""
    array[i], array[j] = array[j], array[i]

# Main Program
print(insertion_sort([1, -12, 9, 54, -11, 75, 8, 14]))

File: C/Sorting/insertion_Sort.c
//Best Case: O(n) Time Complexity | O(1) Space Complexity
//Average Case: O(n^2) Time Complexity | O(1) Space Complexity
//Worst Case: O(n^2) Time Complexity | O(1) Space Complexity

#include<stdio.h>

#define N 8

//Function Signature
int  * insertionSort(int*); //Function to implment insertion sort algorithm
void swap(int*, int*); //Function to swap two elements

int main(){
    int myArray[N] = {1, -12, 9, 54, -11, 75, 8, 14};
    int i = 0;
    int *ptr;

    ptr = insertionSort(myArray);
    for(i = 0; i < N; i++){
        printf(" %d ", *(ptr + i));
    }
    return 0;
}

int * insertionSort(int *myArray){
    int i = 0;
    int lenArray = N;
    int valueSort = 0;

    for(i = 1; i < lenArray; i++){
        //Loop through the array once
        valueSort = *(myArray + i); //Select the first unsorted element
        while(*(myArray + (i - 1)) > valueSort && i > 0){
            swap(&*(myArray + i), &*(myArray + (i - 1))); //Swapping other elements to the right to shift unsorted array
            i = i - 1;
        }
    }
    return myArray;   
}

void swap(int *i, int *j){
    int temp =0; 

    temp = *i;
    *i = *j;
    *j = temp;
}